// playwright-bdd/dist/reporter/cucumber/messagesBuilder/AttachmentMapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttachmentMapper = void 0;
const AutofillMap_js_1 = require("../../../utils/AutofillMap.js");
const pwStepUtils_1 = require("./pwStepUtils");
const stripAnsiEscapes_js_1 = require("../../../utils/stripAnsiEscapes.js");
class AttachmentMapper {
    constructor(result) {
        this.result = result;
        this.stepAttachments = new AutofillMap_js_1.AutofillMap();
        this.unusedAttachments = [];
        this.mapAttachments();
        this.mapUnusedAttachments();
        this.mapStdioAttachments('stdout');
        this.mapStdioAttachments('stderr');
    }
    getStepAttachments(pwStep) {
        return this.stepAttachments.get(pwStep) || [];
    }
    mapAttachments() {
        const allAttachments = this.result.attachments.slice();
        const allAttachmentSteps = (0, pwStepUtils_1.findAllStepsWithCategory)(this.result, 'attach');
        allAttachmentSteps
            // filter out trace/screenshot/video to always show them at the scenario bottom (and not in bg)
            .filter((attachmentStep) => !isPlaywrightAutoAttachment(attachmentStep))
            .forEach((attachmentStep) => this.mapAttachment(attachmentStep, allAttachments));
        this.unusedAttachments.push(...allAttachments);
    }
    mapAttachment(attachmentStep, allAttachments) {
        const index = allAttachments.findIndex((a) => getAttachmentStepTitle(a.name) === attachmentStep.title);
        if (index === -1) {
            throw new Error(`Attachment not found for step: ${attachmentStep.title}`);
        }
        // pick attachment from result.attachments array
        const [foundAttachment] = allAttachments.splice(index, 1);
        const parentStep = attachmentStep.parent;
        const stepAttachments = parentStep
            ? this.stepAttachments.getOrCreate(parentStep, () => [])
            : this.unusedAttachments;
        stepAttachments.push(foundAttachment);
    }
    mapUnusedAttachments() {
        if (!this.unusedAttachments.length)
            return;
        // map unused attachments to the 'After Hooks' step
        const afterHooksRoot = this.getAfterHooksRoot();
        const stepAttachments = this.stepAttachments.getOrCreate(afterHooksRoot, () => []);
        stepAttachments.push(...this.unusedAttachments);
    }
    mapStdioAttachments(name) {
        // map stdout / stderr to the 'After Hooks' step
        if (!this.result[name]?.length)
            return;
        const body = this.result[name]
            // data can be buffer
            .map((data) => (typeof data === 'string' ? data : data?.toString('utf8')))
            .filter(Boolean)
            .map((str) => (0, stripAnsiEscapes_js_1.stripAnsiEscapes)(str))
            .join('');
        const afterHooksRoot = this.getAfterHooksRoot();
        const stepAttachments = this.stepAttachments.getOrCreate(afterHooksRoot, () => []);
        stepAttachments.push({
            name,
            // Attach stdout / stderr as text/x.cucumber.log+plain instead of text/plain,
            // because Cucumber HTML report has pretty formatting for that.
            // See: https://github.com/vitalets/playwright-bdd/issues/239#issuecomment-2451423020
            contentType: 'text/x.cucumber.log+plain',
            body: Buffer.from(body),
        });
    }
    getAfterHooksRoot() {
        const afterHooksRoot = (0, pwStepUtils_1.getHooksRootPwStep)(this.result, 'after');
        if (!afterHooksRoot) {
            throw new Error(`Can not find after hooks root.`);
        }
        return afterHooksRoot;
    }
}
exports.AttachmentMapper = AttachmentMapper;
// See: https://github.com/microsoft/playwright/blob/main/packages/playwright/src/worker/testInfo.ts#L413
function getAttachmentStepTitle(attachmentName) {
    return `attach "${attachmentName}"`;
}
function isPlaywrightAutoAttachment(attachmentStep) {
    return /^attach "(trace|screenshot|video)"$/.test(attachmentStep.title);
}
