// playwright-bdd/dist/reporter/cucumber/messagesBuilder/pwStepUtils.js
"use strict";
/**
 * Utility functions for filtering Playwright steps.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getHooksRootPwStep = getHooksRootPwStep;
exports.findAllStepsWithCategory = findAllStepsWithCategory;
exports.findDeepestStepWithError = findDeepestStepWithError;
exports.findDeepestStepWithUnknownDuration = findDeepestStepWithUnknownDuration;
exports.findAllStepsWith = findAllStepsWith;
exports.isUnknownDuration = isUnknownDuration;
exports.findParent = findParent;
exports.areTestErrorsEqual = areTestErrorsEqual;
exports.isTopLevelStep = isTopLevelStep;
exports.isSkippedError = isSkippedError;
const utils_1 = require("../../../utils");
// Playwright step categories, that can be mapped to testStep / hook in Cucumber messages
const MEANINGFUL_STEP_CATEGORIES = ['hook', 'fixture', 'test.step'];
/**
 * Returns root step for Before Hooks / After Hooks.
 * Strings 'Before Hooks' and 'After Hooks' are hardcoded in Playwright.
 * See: https://github.com/microsoft/playwright/blob/release-1.49/packages/playwright/src/worker/workerMain.ts#L336
 */
function getHooksRootPwStep(result, type) {
    const rootStepTitle = type === 'before' ? 'Before Hooks' : 'After Hooks';
    return result.steps.find((step) => step.category === 'hook' && step.title === rootStepTitle);
}
function findAllStepsWithCategory(result, category) {
    const categories = Array.isArray(category) ? category : [category];
    return findAllStepsWith(result.steps, (pwStep) => categories.includes(pwStep.category));
}
function findDeepestStepWithError(root) {
    if (!root)
        return;
    return findDeepestStepWith(root, (pwStep) => {
        return Boolean(pwStep.error) && MEANINGFUL_STEP_CATEGORIES.includes(pwStep.category);
    });
}
function findDeepestStepWithUnknownDuration(root) {
    if (!root)
        return;
    return findDeepestStepWith(root, (pwStep) => {
        return isUnknownDuration(pwStep) && MEANINGFUL_STEP_CATEGORIES.includes(pwStep.category);
    });
}
/**
 * Finds the deepest step that satisfies predicate function.
 */
function findDeepestStepWith(root, predicate) {
    let result;
    let curSteps = [root];
    while (curSteps.length) {
        const nextSteps = [];
        curSteps.forEach((pwStep) => {
            if (predicate(pwStep))
                result = pwStep;
            nextSteps.push(...(pwStep.steps || []));
        });
        curSteps = nextSteps;
    }
    return result;
}
/**
 * Find all steps that satisfies predicate function.
 */
function findAllStepsWith(root, predicate) {
    const result = [];
    let curSteps = root ? (0, utils_1.toArray)(root) : [];
    while (curSteps.length) {
        const nextSteps = [];
        curSteps.forEach((pwStep) => {
            if (predicate(pwStep))
                result.push(pwStep);
            nextSteps.push(...(pwStep.steps || []));
        });
        curSteps = nextSteps;
    }
    return result;
}
function isUnknownDuration(pwStep) {
    return pwStep.duration === -1;
}
function findParent(pwStep, predicate) {
    let parent = pwStep.parent;
    while (parent) {
        if (predicate(parent))
            return parent;
        parent = parent.parent;
    }
}
function areTestErrorsEqual(e1, e2) {
    // don't check location as it's object
    const keys = ['message', 'stack', 'value'];
    return keys.every((key) => e1[key] === e2[key]);
}
function isTopLevelStep(pwStep) {
    return !pwStep.parent;
}
/**
 * When calling test.skip() in Playwright test, it throws an error with message:
 * "Test is skipped".
 * This error exists in step, but it is not a real error, it is a skipped step.
 * See: https://github.com/microsoft/playwright/blob/main/packages/playwright/src/worker/testInfo.ts#L223
 */
function isSkippedError(error) {
    return Boolean(error?.message?.includes('Test is skipped:'));
}
