// playwright-bdd/dist/reporter/cucumber/messagesBuilder/TestCaseRunHooks.js
"use strict";
/**
 * Executed hooks of test run.
 *
 * In Playwright:
 * worker-level hooks (BeforeAll / AfterAll) are still part of
 * particular test results. E.g. BeforeAll is reported as a step of the first test in a worker.
 *
 * In Cucumber:
 * Worker-level hooks are not considered to be a part of any test case.
 * They will be reported as separate messages TestRunHookStarted / TestRunHookFinished.
 * See: https://github.com/cucumber/messages/pull/102
 *
 * In playwright-bdd:
 * As of now, we don't emit TestRunHookStarted / TestRunHookFinished messages,
 * but include worker-level hooks into testCase steps.
 * This could be changed in the future, when cucumber-js will add it.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestCaseRunHooks = void 0;
const Hook_1 = require("./Hook");
const pwStepUtils_1 = require("./pwStepUtils");
const TestStepRun_1 = require("./TestStepRun");
class TestCaseRunHooks {
    constructor(testCaseRun, hookType, bgSteps) {
        this.testCaseRun = testCaseRun;
        this.hookType = hookType;
        this.bgSteps = bgSteps;
        this.hookPwSteps = new Set();
        this.executedHooks = new Map();
    }
    fill() {
        this.setRootStep();
        this.addStepsWithName();
        this.addStepWithTimeout();
        this.addStepWithError();
        this.addStepWithErrorFallback();
        this.addStepsWithAttachment();
        this.setExecutedHooks();
        return this;
    }
    buildMessages() {
        const messages = [];
        this.testCaseRun
            .getTestCase()
            .getHooks(this.hookType)
            .forEach((hookInfo) => {
            const executedHook = this.executedHooks.get(hookInfo.hook.internalId);
            // todo: if pwStep is not found in this.executedBeforeHooks,
            // it means that this hook comes from another attempt of this test case.
            // We can still try to find it in test result, as otherwise it will be marked as skipped,
            // but actually it was executed.
            const testStepRun = new TestStepRun_1.TestStepRun(this.testCaseRun, hookInfo.testStep, executedHook?.pwStep);
            messages.push(...testStepRun.buildMessages());
        });
        return messages;
    }
    setRootStep() {
        this.rootPwStep = (0, pwStepUtils_1.getHooksRootPwStep)(this.testCaseRun.result, this.hookType);
    }
    addStepsWithName() {
        if (!this.rootPwStep)
            return;
        const stepsWithName = (0, pwStepUtils_1.findAllStepsWith)(this.rootPwStep, (pwStep) => {
            return pwStep.category === 'test.step' && pwStep.title;
        });
        stepsWithName
            .filter((pwStep) => !this.isBackgroundStep(pwStep))
            .forEach((pwStep) => this.hookPwSteps.add(pwStep));
    }
    addStepsWithAttachment() {
        if (!this.rootPwStep)
            return;
        const { attachmentMapper } = this.testCaseRun;
        const stepsWithAttachment = (0, pwStepUtils_1.findAllStepsWith)(this.rootPwStep, (pwStep) => {
            return attachmentMapper.getStepAttachments(pwStep).length > 0;
        });
        stepsWithAttachment
            .filter((pwStep) => !this.isBackgroundStep(pwStep))
            .forEach((pwStep) => this.hookPwSteps.add(pwStep));
    }
    addStepWithTimeout() {
        if (!this.testCaseRun.isTimeouted())
            return;
        if (this.testCaseRun.timeoutedStep)
            return;
        // Search timeouted step by duration = -1.
        // This is not 100% method, sometimes timeouted steps have real duration value.
        // But allows to better place timeout error message in report.
        const timeoutedStep = (0, pwStepUtils_1.findDeepestStepWithUnknownDuration)(this.rootPwStep);
        if (!timeoutedStep || this.isBackgroundStep(timeoutedStep))
            return;
        this.hookPwSteps.add(timeoutedStep);
        this.testCaseRun.registerTimeoutedStep(timeoutedStep);
    }
    addStepWithError() {
        // In case of several errors in after hooks (as they all run),
        // parent pwStep and test result contain only the last error,
        // but each hook step itself contains own error.
        // Here we find only first step with error.
        // Todo: find and show all errors for after hooks.
        const stepWithError = (0, pwStepUtils_1.findDeepestStepWithError)(this.rootPwStep);
        if (!stepWithError || this.isBackgroundStep(stepWithError))
            return;
        // If step is already added to errorSteps, don't register it as hookStep.
        // This is mainly for timeout steps in hooks or bg:
        // They have duration -1 and no 'error' field, but their parent has 'error' field.
        // Here we find this parent again and avoid reporting the error twice.
        if (!this.testCaseRun.hasRegisteredError(stepWithError.error)) {
            this.registerErrorStep(stepWithError, stepWithError.error);
        }
    }
    // eslint-disable-next-line visual/complexity
    addStepWithErrorFallback() {
        // if in 'after' hooks group there are unprocessed errors,
        // attach them to After Hooks root step.
        if (this.hookType !== 'after')
            return;
        const unprocessedErrors = this.testCaseRun.getUnprocessedErrors();
        if (unprocessedErrors.length === 0)
            return;
        const error = buildFallbackError(unprocessedErrors);
        // if there is timeouted step without attached error, attach all unprocessed errors to it,
        // otherwise attach unprocessed error to root 'After Hooks' step.
        const { timeoutedStep } = this.testCaseRun;
        if (timeoutedStep && !this.testCaseRun.getStepError(timeoutedStep)) {
            this.testCaseRun.registerErrorStep(timeoutedStep, error);
        }
        else if (this.rootPwStep) {
            this.registerErrorStep(this.rootPwStep, error);
        }
    }
    setExecutedHooks() {
        this.hookPwSteps.forEach((pwStep) => {
            const internalId = Hook_1.Hook.getInternalId(pwStep);
            const hook = this.getOrRegisterHook(pwStep);
            this.executedHooks.set(internalId, { hook, pwStep });
        });
    }
    getOrRegisterHook(pwStep) {
        const internalId = Hook_1.Hook.getInternalId(pwStep);
        return this.testCaseRun.hooks.getOrCreate(internalId, () => new Hook_1.Hook(internalId, this.hookType, pwStep));
    }
    registerErrorStep(pwStep, error) {
        this.hookPwSteps.add(pwStep);
        this.testCaseRun.registerErrorStep(pwStep, error);
    }
    isBackgroundStep(pwStep) {
        return this.bgSteps.has(pwStep);
    }
}
exports.TestCaseRunHooks = TestCaseRunHooks;
function buildFallbackError(unprocessedErrors) {
    return unprocessedErrors.length === 1
        ? unprocessedErrors[0]
        : buildConcatenatedError(unprocessedErrors);
}
function buildConcatenatedError(errors) {
    const message = errors
        .map((e) => e.message)
        .filter(Boolean)
        .join('\n\n');
    return { message };
}
